Support for XPath functions

We need support for XPath functions to evaluate XPath expressions like the following:
```
substring(/TEI/teiHeader/fileDesc/notesStmt/note[@type="digDate"]/date[@ana="#scan"]/@when,1,10)
```

Currently, an Exception is thrown:
```
[...]
Caused by: javax.xml.xpath.XPathExpressionException: net.sf.saxon.trans.XPathException: Cannot convert XPath value to Java object: required class is org.w3c.dom.NodeList; supplied value has type xs:string
	at net.sf.saxon.xpath.XPathExpressionImpl.evaluate(XPathExpressionImpl.java:244)
	at de.digitalcollections.commons.xml.xpath.XPathWrapper.evaluateXpath(XPathWrapper.java:187)
	... 126 more
Caused by: net.sf.saxon.trans.XPathException: Cannot convert XPath value to Java object: required class is org.w3c.dom.NodeList; supplied value has type xs:string
	at net.sf.saxon.dom.DOMObjectModel.convertXPathValueToObject(DOMObjectModel.java:420)
	at net.sf.saxon.dom.DOMObjectModel$7.convert(DOMObjectModel.java:202)
	at net.sf.saxon.xpath.XPathExpressionImpl.evaluate(XPathExpressionImpl.java:239)
``

--- a/dc-commons-xml/pom.xml
+++ b/dc-commons-xml/pom.xml
@@ -10,7 +10,7 @@
 
   <name>DigitalCollections: Commons XML</name>
   <artifactId>dc-commons-xml</artifactId>
-  <version>5.1.2-SNAPSHOT</version>
+  <version>5.1.2</version>
   <packaging>jar</packaging>
   
   <properties>

--- a/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
+++ b/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
@@ -7,7 +7,6 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
-import java.util.LinkedList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -301,25 +300,31 @@ class DocumentReader {
     }
   }
 
+  // TODO Don't restrict to string lists
   private List<String> resolveVariablesAsStrings(String[] paths, boolean multiValued) {
-    List<String> result = new LinkedList<>();
+    List<String> result = new ArrayList<>();
     paths = prependWithRootPaths(paths);
 
     for (String path : paths) {
       for (Object resolvedObject : xpw.asListOfObjects(path)) {
+
         if (resolvedObject instanceof String && !((String) resolvedObject).isEmpty()) {
           result.add((String) resolvedObject);
+          continue;
         }
-        if (multiValued && resolvedObject instanceof DOMNodeList) {
+        if (resolvedObject instanceof Integer) {
+          continue;
+        }
+        if (resolvedObject instanceof DOMNodeList) {
           DOMNodeList nodeList = ((DOMNodeList) resolvedObject);
           for (int i = 0, l = nodeList.getLength(); i < l; i++) {
             String textContent = nodeList.item(i).getTextContent();
             if (textContent == null) {
               textContent = "";
             }
-            textContent = textContent.trim();
-            if (!result.contains(textContent)) {
-              result.add(textContent);
+            result.add(textContent.trim());
+            if (!multiValued) {
+              break;
             }
           }
         }

--- a/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
+++ b/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
@@ -313,6 +313,7 @@ class DocumentReader {
           continue;
         }
         if (resolvedObject instanceof Integer) {
+          result.add(((Integer) resolvedObject).toString());
           continue;
         }
         if (resolvedObject instanceof DOMNodeList) {

--- a/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
+++ b/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/DocumentReader.java
@@ -7,6 +7,7 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -16,6 +17,7 @@ import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import javax.xml.xpath.XPathExpressionException;
+import net.sf.saxon.dom.DOMNodeList;
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
 import org.w3c.dom.Node;
@@ -40,25 +42,23 @@ class DocumentReader {
   }
 
   public String readValue(List<String> expressions) throws XPathMappingException {
-    return resolveVariable(expressions.toArray(new String[] {})).values().stream()
+    return resolveVariablesAsStrings(expressions.toArray(new String[] {}), false).stream()
         .findFirst()
         .orElse(null);
   }
 
   public List<String> readValues(List<String> expressions) throws XPathMappingException {
-    return readLocalizedValues(expressions).values().stream()
-        .flatMap(List::stream)
-        .collect(Collectors.toList());
+    return resolveVariablesAsStrings(expressions.toArray(new String[] {}), true);
   }
 
   public Map<Locale, String> readLocalizedValue(List<String> expressions)
       throws XPathMappingException {
-    return resolveVariable(expressions.toArray(new String[] {}));
+    return resolveLocalizedVariable(expressions.toArray(new String[] {}));
   }
 
   public Map<Locale, List<String>> readLocalizedValues(List<String> expressions)
       throws XPathMappingException {
-    return resolveVariable(expressions.toArray(new String[] {}), true);
+    return resolveLocalizedVariable(expressions.toArray(new String[] {}), true);
   }
 
   public List<Element> readElementList(List<String> expressions) throws XPathMappingException {
@@ -87,7 +87,7 @@ class DocumentReader {
     Map<String, Map<Locale, String>> resolvedVariables = new HashMap<>();
     for (String requiredVariable : requiredVariables) {
       XPathVariable var = givenVariablesByName.get(requiredVariable);
-      resolvedVariables.put(var.name(), this.resolveVariable(var.paths()));
+      resolvedVariables.put(var.name(), this.resolveLocalizedVariable(var.paths()));
     }
     return this.executeTemplate(template, resolvedVariables);
   }
@@ -213,12 +213,13 @@ class DocumentReader {
     }
   }
 
-  private Map<Locale, String> resolveVariable(String[] paths) throws XPathMappingException {
-    return resolveVariable(paths, false).entrySet().stream()
+  private Map<Locale, String> resolveLocalizedVariable(String[] paths)
+      throws XPathMappingException {
+    return resolveLocalizedVariable(paths, false).entrySet().stream()
         .collect(Collectors.toMap(Entry::getKey, var -> var.getValue().get(0)));
   }
 
-  private Map<Locale, List<String>> resolveVariable(String[] paths, boolean multiValued)
+  private Map<Locale, List<String>> resolveLocalizedVariable(String[] paths, boolean multiValued)
       throws XPathMappingException {
     paths = prependWithRootPaths(paths);
     Map<Locale, LinkedHashSet<String>> result = new LinkedHashMap<>();
@@ -298,6 +299,34 @@ class DocumentReader {
     } catch (IllegalArgumentException e) {
       throw new XPathMappingException("Failed to resolve XPath", e);
     }
+  }
+
+  private List<String> resolveVariablesAsStrings(String[] paths, boolean multiValued) {
+    List<String> result = new LinkedList<>();
+    paths = prependWithRootPaths(paths);
+
+    for (String path : paths) {
+      for (Object resolvedObject : xpw.asListOfObjects(path)) {
+        if (resolvedObject instanceof String && !((String) resolvedObject).isEmpty()) {
+          result.add((String) resolvedObject);
+        }
+        if (multiValued && resolvedObject instanceof DOMNodeList) {
+          DOMNodeList nodeList = ((DOMNodeList) resolvedObject);
+          for (int i = 0, l = nodeList.getLength(); i < l; i++) {
+            String textContent = nodeList.item(i).getTextContent();
+            if (textContent == null) {
+              textContent = "";
+            }
+            textContent = textContent.trim();
+            if (!result.contains(textContent)) {
+              result.add(textContent);
+            }
+          }
+        }
+      }
+    }
+
+    return result;
   }
 
   protected static Locale determineLocaleFromCode(String localeCode) {

--- a/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathWrapper.java
+++ b/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathWrapper.java
@@ -4,6 +4,7 @@ import de.digitalcollections.commons.xml.namespaces.DigitalCollectionsNamespaceC
 import java.io.StringWriter;
 import java.io.UnsupportedEncodingException;
 import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Stack;
 import javax.xml.namespace.NamespaceContext;
@@ -109,6 +110,18 @@ public class XPathWrapper {
       list.add(nodeList.item(i));
     }
     return list;
+  }
+
+  public List<Object> asListOfObjects(String xpath) {
+    List<Object> ret = new LinkedList<>();
+    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.STRING));
+    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NUMBER));
+    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.BOOLEAN));
+    try {
+      ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NODESET));
+    } catch (Exception ignore) {
+    }
+    return ret;
   }
 
   /**

--- a/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathWrapper.java
+++ b/dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathWrapper.java
@@ -112,14 +112,27 @@ public class XPathWrapper {
     return list;
   }
 
+  /**
+   * Evaluate every possible type, an XPathExpression can return
+   *
+   * @param xpath the xPath expression
+   * @return all possible xPath evaluations as a list of Objects
+   */
   public List<Object> asListOfObjects(String xpath) {
     List<Object> ret = new LinkedList<>();
-    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.STRING));
-    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NUMBER));
-    ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.BOOLEAN));
     try {
       ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NODESET));
-    } catch (Exception ignore) {
+    } catch (Exception noNodeset) {
+      // We have no NodeSet as return type, so we continue with the "flat" types
+      ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.STRING));
+      ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NUMBER));
+      ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.BOOLEAN));
+      // ... and finally, we check, if the evaluation returns a single XPath node
+      try {
+        ret.add(evaluateXpath(getDocument(), xpath, XPathConstants.NODE));
+      } catch (Exception noNode) {
+        // to be ignored
+      }
     }
     return ret;
   }

--- a/dc-commons-xml/src/test/java/de/digitalcollections/commons/xml/xpath/XPathMapperTest.java
+++ b/dc-commons-xml/src/test/java/de/digitalcollections/commons/xml/xpath/XPathMapperTest.java
@@ -253,6 +253,23 @@ public class XPathMapperTest {
     assertThat(outerMapper.getEmptyXPathRootInnerRoot()).isNotNull();
     assertThat(outerMapper.getEmptyXPathRootInnerRoot().getXmlId()).isEqualTo("bsb00050852");
   }
+
+  @DisplayName(
+      "shall be able to evaluate statements, which return integer values by returning their string representation")
+  @Test
+  public void testStatementsWithIntegerResult() throws XPathMappingException {
+    TestMapper mapper = testMapperFixture.setUpMapperWithResource("bsbstruc.xml");
+    assertThat(mapper.getAmountPlaces()).isEqualTo(3);
+  }
+
+  @DisplayName(
+      "shall be able to evaluate statements, which return a node by returning its string representation")
+  @Test
+  public void testStatementsWithReturnNode() throws XPathMappingException {
+    TestMapper mapper = testMapperFixture.setUpMapperWithResource("bsbstruc.xml");
+    assertThat(mapper.getFirstPersNameNode()).contains("Kugelmann, Hans");
+    assertThat(mapper.getFirstPersNameNode()).contains("Name, English");
+  }
   // ---------------------------------------------------------------------------------------------
 
   @XPathRoot(defaultNamespace = "http://www.tei-c.org/ns/1.0")
@@ -379,6 +396,24 @@ public class XPathMapperTest {
 
     public String getDateScan() {
       return dateScan;
+    }
+
+    @XPathBinding("count(" + BIBLSTRUCT_PATH + "/monogr/imprint/pubPlace)")
+    void setAmountPlaces(String strAmountPlaces) {
+      this.amountPlaces = Integer.parseInt(strAmountPlaces);
+    }
+
+    int amountPlaces;
+
+    public int getAmountPlaces() {
+      return amountPlaces;
+    }
+
+    @XPathBinding(BIBLSTRUCT_PATH + "/monogr/author/persName[1]")
+    String firstPersNameNode;
+
+    public String getFirstPersNameNode() {
+      return firstPersNameNode;
     }
   }
 

--- a/dc-commons-xml/src/test/java/de/digitalcollections/commons/xml/xpath/XPathMapperTest.java
+++ b/dc-commons-xml/src/test/java/de/digitalcollections/commons/xml/xpath/XPathMapperTest.java
@@ -92,6 +92,13 @@ public class XPathMapperTest {
     assertThat(mapper.getFirstPlace()).isEqualTo("Augsburg");
   }
 
+  @DisplayName("shall evaluate a single valued expression with a function")
+  @Test
+  public void testSingleValueExpressionWithFunction() throws Exception {
+    TestMapper mapper = testMapperFixture.setUpMapperWithResource("bsbstruc.xml");
+    assertThat(mapper.getDateScan()).isEqualTo("2019-11-19");
+  }
+
   @DisplayName("shall return multivalued contents in the same order as in the bind")
   @Test
   public void testMultivaluedFieldsAndTheirOrder() throws Exception {
@@ -364,6 +371,14 @@ public class XPathMapperTest {
 
     String getNoPlace() {
       return noPlace;
+    }
+
+    @XPathBinding(
+        "substring(/TEI/teiHeader/fileDesc/notesStmt/note[@type=\"digDate\"]/date[@ana=\"#scan\"]/@when,1,10)")
+    String dateScan;
+
+    public String getDateScan() {
+      return dateScan;
     }
   }
 

--- a/dc-commons-xml/src/test/resources/bsbstruc.xml
+++ b/dc-commons-xml/src/test/resources/bsbstruc.xml
@@ -25,6 +25,11 @@
           </p>
         </availability>
       </publicationStmt>
+      <notesStmt>
+        <note type="digDate">
+          <date when="2019-11-19T20:51:49" ana="#scan"/>
+        </note>
+      </notesStmt>
       <sourceDesc>
         <listBibl type="source">
           <biblStruct ana="#bsbStandard">

